src/Lecture2.hs

```diff
@@ -288,4 +364,23 @@ Write a function that takes and expression and performs "Constant
 Folding" optimization on the given expression.
 -}
 constantFolding :: Expr -> Expr
-constantFolding = error "TODO"
+constantFolding = addExpr . extractConstant
+    where
+      extractConstant :: Expr -> ([Int], [Expr])
+      extractConstant (Lit i)   = ([i], [])
+      extractConstant (Var x)   = ([], [Var x])
+      extractConstant (Add x y) = (x1 ++ y1, x2 ++ y2)
+        where
+          (x1, x2) = extractConstant x
+          (y1, y2) = extractConstant y
+      addExpr :: ([Int], [Expr]) -> Expr
+      addExpr (c, e) = case (c, e) of
+        ([], []) -> error "invalid"
+        (xs, []) -> Lit (sum xs)
+        ([], ys) -> addVars ys
+        (xs, ys) -> if sum xs == 0 then addVars ys else Add (Lit (sum xs)) (addVars ys)
+      addVars :: [Expr] -> Expr
+      addVars []             = error "Invalid"
+      addVars [x]            = x
+      addVars [x, y]         = Add x y
+      addVars (x : y : rest) = Add (Add x y) (addVars rest)
```

Comment:
You can get rid of handling this impossible case too. In the pair you already pattern matched on the list, so you know that it's not empty. So yo can pass it's first element and the rest to the function and continue doing this recursively to avoid throwing `error`

```suggestion
        ([], y : ys) -> addVars y ys
        (xs, ys) -> if sum xs == 0 then addVars ys else Add (Lit (sum xs)) (addVars ys)
      addVars :: Expr -> [Expr] -> Expr
      addVars e [] = e
      addVars e (x : xs) = ...
```

Reply:
Thanks @chshersh , will need to think about this some more.